### PR TITLE
WIP: Rewrite examples with `async`.

### PR DESCRIPTION
**Pull request by rubdos**
## Motivation

> We would love to get community involvement in this work. There is a lot to do. The best way to get involved is to:

> The examples should be updated to idiomatic async / await syntax.

> The first alpha release will happen once the examples have been updated.

## Solution

I started rewriting the examples using the new `#[tokio::main]` and `async` and `await` keywords. I will have every example in it's own commit, such that you can cherry pick what you'd like to keep.

I'm on Gitter and IRC, so feel free to ping me there.

### CODE REVIEW

File: tokio/examples/echo-udp.rs [hunk 1]
```diff
@@ -10,44 +10,47 @@
 //!
 //! Each line you type in to the `nc` terminal should be echo'd back to you!
 
-#![deny(warnings, rust_2018_idioms)]
+#![feature(async_await)]
+// #![deny(warnings, rust_2018_idioms)]
```
Comment (jonhoo) [inline tokio/examples/echo-udp.rs]: Any particular reason why you commented out this `deny`?

Comment (rubdos) [reply tokio/examples/echo-udp.rs]: Yes, I was playing around. I'll put it back :-)

**Comment by rubdos** [comment]
I think these three examples are as far as it gets right now:

- [ ] Anything TCP basically depends on Streams (i.e., `incoming` feature for `tokio-tcp`), which is [to be decided](https://github.com/tokio-rs/tokio/issues/1225). Without the `incoming` feature, you need to manually use the `poll_accept` method, which is quite counter productive.
- [ ] Codecs still seem to be broken, and they also depend on Streams and Sinks.

I'll still have a look at `blocking.rs` and `manual_runtime.rs`, but especially the latter is not my personal use case, so I might have no clue :-)

Everything else seems to fall in the above categories.

**Comment by rubdos** [comment]
`blocking.rs` depends on a threadpool's `block_on_all`, which is currently [still commented out](https://github.com/tokio-rs/tokio/issues/1200).

**Comment by rubdos** [comment]
`manual_runtime.rs` depends on [timer support](https://github.com/tokio-rs/tokio/pull/1222), but that's soon there.

**Comment by carllerche** [comment]
Thanks for doing this work.

re: `incoming`, I would rewrite the examples to not use streaming (which works better with async / await). For example:

```rust
let listener = TcpListener::bind(...)?;

loop {
    let (socket, _) = listener.accept().await?;
    // do something with the socket.
}
```

I'll review the rest shortly. Thanks again.

**Comment by sopium** [comment]
I updated the proxy example in the TCP split PR: [link](https://github.com/tokio-rs/tokio/pull/1217#issuecomment-506894355), but did not include it eventually. The usage of `incoming` can be replaced with `accept`. But it also needs `try_join`.

**Comment by rubdos** [comment]
> I updated the proxy example in the TCP split PR: [link](https://github.com/tokio-rs/tokio/pull/1217#issuecomment-506894355), but did not include it eventually.

Mind recovering that commit and filing a merge request on my merge request? I'm not sure whether I can pull that commit of GH.

**Comment by rubdos** [comment]
Hmm, I have the commit now, but there's a lot of things going on that I'm afraid to touch without permission for `proxy`:

- `tokio-tcp` requires `default = ["incoming"]`
- `tokio` requires `futures-preview = "0.3.0-alpha.16"`

**Comment by rubdos** [comment]
> Thanks for doing this work.
> 
> re: `incoming`, I would rewrite the examples to not use streaming (which works better with async / await). For example:
> 
> ```rust
> let listener = TcpListener::bind(...)?;
> 
> loop {
>     let (socket, _) = listener.accept().await?;
>     // do something with the socket.
> }
> ```
> 
> I'll review the rest shortly. Thanks again.

You didn't mean to rewrite the Tcp examples with `fn accept()`, right? Because I cannot find any `accept()` :-)

```
error[E0599]: no method named `accept` found for type `tokio_tcp::listener::TcpListener` in the current scope
  --> tokio/examples/proxy.rs:48:29
   |
48 |         let stream = socket.accept().await.unwrap()?;
   |                             ^^^^^^
```

**Comment by carllerche** [comment]
@rubdos You are correct! I have a branch in progress, but I ended up going into a yak shave as the necessary tooling to write tests weren't in place yet. I will try to get that up today.

**Comment by sopium** [comment]
> 
> 
> Hmm, I have the commit now, but there's a lot of things going on that I'm afraid to touch without permission for `proxy`:
> 
>     * `tokio-tcp` requires `default = ["incoming"]`
> 
>     * `tokio` requires `futures-preview = "0.3.0-alpha.16"`
> 
> 
> EDIT: cfr commit [4487254](https://github.com/tokio-rs/tokio/commit/4487254ddf0206898fc4ae28157d358f59534257)

Yeah, maybe it's not time yet. It can be adopted once the incoming/accept method is settled, and `try_join` is implemented or allowed to be imported from futures/futures-utils.

Also, feel free to come up with something entirely different. It's by no means the only way to write this example.

**Comment by carllerche** [comment]
I have posted #1242.

**Comment by rubdos** [comment]
Nice! Expect TCP examples coming in tonight!

**Pull request merged**